src/Interpreters/HashJoin/AddedColumns.h

```diff
@@ -115,25 +113,81 @@ class AddedColumns
             if (columns[j]->isNullable() && !saved_column->isNullable())
                 nullable_column_ptrs[j] = typeid_cast<ColumnNullable *>(columns[j].get());
         }
+        join_data_avg_perkey_rows = join.getJoinedData()->avgPerKeyRows();
+        output_by_row_list_threshold = join.getTableJoin().outputByRowListPerkeyRowsThreshold();
     }
 
     size_t size() const { return columns.size(); }
 
-    void buildOutput();
+    void buildOutputFromRowRef();
+
+    void buildOutputFromRowRefList();
+
+    void buildJoinGetOutput();
+
+    template<bool from_row_list>
+    void buildOutputFromBlocks()
+    {
+        if (this->size() == 0)
+            return;
+        std::vector<const Block *> blocks;
+        std::vector<UInt32> row_nums;
+        blocks.reserve(lazy_output.row_refs.size());
+        row_nums.reserve(lazy_output.row_refs.size());
+        for (auto row_ref_i : lazy_output.row_refs)
```

Comment:
pls leave a comment explaining why do we do this. it is really tricky place

Reply:
done